### DISCUSSION TITLE
target range for textDocument/definition

### DISCUSSION THREAD

**Discussion by damnskippy** [clangd/clangd#2274] [Q&A] [answered]
Hello, in the results for textDocument/definition it seems the range has the start and end line be the same value i.e. the line where the said definition is anchored.  ccls for instance returns the targetRange which contains the start and end lines of the definition.  This makes it possible to highlight, or display the definition in a new popup window, etc.
Not 100% sure but this seems sort of related: pull request on this subject: https://github.com/llvm/llvm-project/pull/118102
linked to https://github.com/clangd/clangd/issues/59
Thanks.

**Comment by HighCommander4** [comment, marked as answer]
To add some more details here: the [protocol](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_definition) allows the response to `textDocument/definition` to be either `Location[]` or `LocationLink[]`, the latter being newly added in version 3.14 of the protocol.

Clangd currently uses the older `Location[]` format, where each result has just a single range (in which case that range pointing to the symbol name token in the definition is appropriate).

I agree that it would make a nice enhancement to upgrade clangd to use the newer `LocationLink[]` format, where each result includes both a "selection range" (narrower range pointing to just the name) and the full range of the symbol definition.

You are right to identify https://github.com/llvm/llvm-project/pull/118102 as a prerequisite for implementing this enhancement.

**Comment by damnskippy** [comment]
Thank you for the clarification and the additional context, very helpful.  Indeed with ccls I see the response is very much as you described.  Please advise if I should open a new issue for this.

**Reply by kadircet** [reply]
can you elaborate on your use case as well? what functionality would improve if clangd were to provide location links?

**Reply by HighCommander4** [reply]
> can you elaborate on your use case as well? what functionality would improve if clangd were to provide location links?

The original post gives a couple of use cases:

> This makes it possible to highlight, or display the definition in a new popup window

I'm assuming "highlight" here means, if the definition is in view in the same file or in another open editor, to highlight its full source range. That sounds like a fairly narrow use case, however "display the definition in a popup window" seems pretty widely applicable and useful.

**Reply by damnskippy** [reply]
Yes, that's accurate.

**Comment by damnskippy** [comment]
Hello, just checking in to see if there's been any consideration to this functionality.

**Reply by HighCommander4** [reply]
Based on the discussion in https://github.com/llvm/llvm-project/pull/118102, it seems like implementing this would require an increase to clangd's memory footprint that's hard to justify for the benefits it brings.

I think progress on this would require either an alternative implementation approach which has a significantly smaller impact on clangd's memory footprint, or other new information that would change the analysis of the tradeoffs (e.g. showing that this would enable features that are supported by many clients and would bring users benefits they can't get another way).

**Reply by damnskippy** [reply]
Understood, thanks.
